analytics/transform.go

```diff
@@ -0,0 +1,270 @@
+//
+// Copyright (c) 2021 Snowplow Analytics Ltd. All rights reserved.
+//
+// This program is licensed to you under the Apache License Version 2.0,
+// and you may not use this file except in compliance with the Apache License Version 2.0.
+// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the Apache License Version 2.0 is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+//
+
+package analytics
+
+import (
+	"encoding/json" // TODO: Look into faster options: https://github.com/json-iterator/go-benchmark https://github.com/buger/jsonparser
+	"fmt"
+	"github.com/pkg/errors"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type KeyVal struct {
+	Key   string
+	Value interface{}
+}
+
+type ValueParser func(string, string) ([]KeyVal, error)
+
+type KeyFunctionPair struct {
+	Key           string
+	ParseFunction ValueParser
+}
+
+// Using a pointer as one can't return a nil value for time.Time, only a zero value - which might be prone to issues in usage of the sdk.
+// Unsure if this is the best decision
+func parseNullableTime(timeString string) (*time.Time, error) {
```

Comment:
Yup, good points both. I think I started with pointer for some reason that then became unimportant.

Reply:
I've removed `parseNullableTime` and moved the code into `parseTime` - which still returns `nil, err` if it hits an error. I don't think this poses any issue, since the public methods which use it will all return `nil, err` if any of the parser functions hit an error, regardless of the parser's other return value.